python intermediate/hero_game/classes6.py

```diff
@@ -0,0 +1,193 @@
+    
+    def is_defend(self) -> None:
+        """
+        activates the defend function
+        """
+        self.defend = True
+
+    
+    def reduce_health(self, monster) -> float:
+        """
+        the reduce health function
+        """
+        current_health = self.hp
+        diff = current_health - (self.hp * 1.8) # help the logics of the game
+
+        if self.defend:
+            self.hp = self.hp * 1.8
+
+        self.hp -= monster.damage
+
+        if self.hp < 0:
+            self.hp = 0
+
+        if self.defend:
+            if diff > monster.damage:
+                self.hp = current_health
+            else: 
+                self.hp -= (diff - monster.damage)
+
+            self.defend = False
+        
+        return self.hp
+
+    
+    def increase_coins(self, coins: int) -> None:
+        """
+        increases coins as the amount of coins
+        """
+        self.coins += coins
+
+    
+    def __str__(self) -> str:
+        """
+        """
+        return f"hero: level - {self.level}, hp - {self.hp}, damage - {self.damage}"
+
+
+class Monster:
+    def __init__(self, name: str, level_hero: int) -> object:
+        generate_level = random.randint(level_hero - 1, level_hero + 1)
+
+        if generate_level == 0:
+            generate_level = 1
+        
+        self.name = name 
+        self.level = generate_level
+        self.hp = generate_level * PROPORTION
+        self.damage = generate_level * PROPORTION
+
+    
+    def attack(self, hero: Hero) -> None:
+        """
+        attack the hero
+        """
+        hero.reduce_health(self)
+
+    
+    def reduce_health(self, hero: Hero) -> float:
+        """
+        reduces the health of the hero 
+        """
+        self.hp -= hero.damage
+
+        if self.hp < 0:
+            self.hp = 0
+
+        return self.hp
+    
+
+    def __str__(self) -> str:
+        return f"monster: level - {self.level}, hp - {self.hp}, damage - {self.damage}"
+    
+
+def choose_action(hero: Hero, monster: Monster) -> float:
+    """
+    choose action of the player
+    input: hero and monster
+    """
+    print("Choose an action")
+    print("(1) attack")
+    print("(2) level up")
+    print("(3) regain health")
+    print("(4) defend")
+    selection = input("enter selection: ")
+
+    while selection != "1" and selection != "2" and selection != "3" and selection != "4":
+        selection = input("enter selection: ")
+
+    if selection == "1":
+        return monster.reduce_health(hero)
+
+    if selection == "2":
+        if not hero.level_up():
+            print("you cant, choose other other action")
+            choose_action(hero, monster)
+
+    if selection == "3":
+        hero.heal()
+
+    if selection == "4":
+        hero.is_defend()
+    
+    hero.coins += 1
+    return -1
+
+
+def main_function() -> None:
+    """
+    the main logic of the game
+    """
+    my_hero = Hero(input("enter name of your hero"))
+    counter = 1
+
+    while True: 
```

Comment:
I still think its better to avoid `while True` especially when it is easy to put there a condition :)